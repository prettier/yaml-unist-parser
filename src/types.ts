import type * as YAML from "yaml";

export type ParsedCST = ReturnType<typeof YAML.parseCST>;

export interface Node {
  type: string;
  position: Position;
  /** @internal non-enumerable */
  _parent?: YamlUnistNode | null;
}

export interface Position {
  start: Point;
  end: Point;
}

export interface Range {
  origStart: number;
  origEnd: number;
}

export interface Point {
  /** 1-based */
  line: number;
  /** 1-based */
  column: number;
  /** 0-based */
  offset: number;
}

export interface Parent extends Node {
  children: Node[];
}

export interface Literal extends Node {
  value: string;
}

// -----------------------------------------------------------------------------

export interface Content {
  anchor: null | Anchor;
  tag: null | Tag;
  /** comments between the node and its tag/anchor */
  middleComments: Comment[];
}

export interface CommentAttachable
  extends LeadingCommentAttachable,
    TrailingCommentAttachable {}

export interface LeadingCommentAttachable {
  /** comments in front of the node */
  leadingComments: Comment[];
}

export interface TrailingCommentAttachable {
  /** comment on the same line of the node */
  trailingComment: null | Comment;
}

export interface EndCommentAttachable {
  /** comments after the node with greater column */
  endComments: Comment[];
}

export interface YAMLSyntaxError extends SyntaxError {
  source: string;
  position: Position;
}

export type YamlUnistNode =
  | Comment
  | Tag
  | Anchor
  | Root
  | Document
  | DocumentHead
  | DocumentBody
  | Directive
  | Alias
  | BlockLiteral
  | BlockFolded
  | Plain
  | QuoteSingle
  | QuoteDouble
  | Mapping
  | MappingItem
  | MappingKey
  | MappingValue
  | Sequence
  | SequenceItem
  | FlowMapping
  | FlowMappingItem
  | FlowSequence
  | FlowSequenceItem;

export type ContentNode = Extract<YamlUnistNode, Content>;

// -----------------------------------------------------------------------------

export interface Comment extends Literal {
  type: "comment";
}

export interface Anchor extends Literal {
  type: "anchor";
}

export interface Tag extends Literal {
  type: "tag";
}

export interface Root extends Parent {
  type: "root";
  children: Document[];
  comments: Comment[];
}

export interface Document extends Parent, TrailingCommentAttachable {
  type: "document";
  children: [DocumentHead, DocumentBody];
}

export interface DocumentHead
  extends Parent,
    EndCommentAttachable,
    TrailingCommentAttachable {
  type: "documentHead";
  children: Directive[];
}

export interface DocumentBody extends Parent, EndCommentAttachable {
  type: "documentBody";
  children: [] | [ContentNode];
}

export interface Directive extends Node, CommentAttachable {
  type: "directive";
  name: string;
  parameters: string[];
}

export interface Alias extends Literal, Content, CommentAttachable {
  type: "alias";
}

export interface BlockValue extends Literal, Content, LeadingCommentAttachable {
  chomping: "clip" | "keep" | "strip";
  indent: null | number;
  /** comment between indicator and the value */
  indicatorComment: null | Comment;
}

export interface BlockLiteral extends BlockValue {
  type: "blockLiteral";
}

export interface BlockFolded extends BlockValue {
  type: "blockFolded";
}

export interface Plain extends Literal, Content, CommentAttachable {
  type: "plain";
}

export interface QuoteValue extends Literal, Content, CommentAttachable {}

export interface QuoteSingle extends QuoteValue {
  type: "quoteSingle";
}

export interface QuoteDouble extends QuoteValue {
  type: "quoteDouble";
}

export interface Mapping extends Parent, Content, LeadingCommentAttachable {
  type: "mapping";
  children: MappingItem[];
}

export interface MappingItemBase extends Parent, LeadingCommentAttachable {
  /** key-value pair */
  children: [MappingKey, MappingValue];
}

export interface MappingItem extends MappingItemBase {
  type: "mappingItem";
}

export interface MappingKey
  extends Parent,
    TrailingCommentAttachable,
    EndCommentAttachable {
  type: "mappingKey";
  children: [] | [ContentNode];
}

export interface MappingValue
  extends Parent,
    CommentAttachable,
    EndCommentAttachable {
  type: "mappingValue";
  children: [] | [ContentNode];
}

export interface Sequence
  extends Parent,
    Content,
    LeadingCommentAttachable,
    EndCommentAttachable {
  type: "sequence";
  children: SequenceItem[];
}

export interface SequenceItemBase extends Parent {
  children: [] | [ContentNode];
}

export interface SequenceItem
  extends SequenceItemBase,
    CommentAttachable,
    EndCommentAttachable {
  type: "sequenceItem";
}

export interface FlowCollection
  extends Parent,
    Content,
    CommentAttachable,
    EndCommentAttachable {
  children: Array<FlowMappingItem | FlowSequenceItem>;
}

export interface FlowMapping extends FlowCollection {
  type: "flowMapping";
  children: FlowMappingItem[];
}

export interface FlowMappingItem extends MappingItemBase {
  type: "flowMappingItem";
}

export interface FlowSequence extends FlowCollection {
  type: "flowSequence";
  children: Array<FlowMappingItem | FlowSequenceItem>;
}

export interface FlowSequenceItem extends SequenceItemBase {
  type: "flowSequenceItem";
}
