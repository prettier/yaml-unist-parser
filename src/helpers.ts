import { wrap } from "jest-snapshot-serializer-raw";
import { parse } from "./parse";
import {
  Anchor,
  Comment,
  Node,
  Position,
  Root,
  Tag,
  YamlUnistNode,
} from "./types";
import * as YAML from "./yaml";

export type Arrayable<T> = T | T[];

export type TestCase = TestCaseSingle | TestCaseMulti;
export type TestCaseSelector = (root: Root) => Arrayable<YamlUnistNode>;

export type TestCaseSingle = [string, TestCaseSelector];
export type TestCaseMulti = [string, TestCaseSelector[]];

export function getFirstContent<T extends YamlUnistNode>(): (root: Root) => T;
export function getFirstContent<T extends YamlUnistNode>(root: Root): T;
export function getFirstContent<T extends YamlUnistNode>(root?: Root) {
  return !root
    ? (_root: Root) => _root.children[0].children[1].children[0] as T
    : (root.children[0].children[1].children[0] as T);
}

export interface SnapshotNodeOptions extends StringifyNodeOptions {
  codeFrameMaxHeight?: number;
  selectNodeToStringify?: (node: YamlUnistNode) => YamlUnistNode;
}

export function testCases(
  cases: TestCase[],
  options: SnapshotNodeOptions = {},
) {
  cases.forEach(testCase => {
    const [text, selector] = testCase;
    const root = parse(text);

    testCrLf(root, text);

    const selectNodes = ([] as TestCaseSelector[]).concat(selector);
    selectNodes.forEach(selectNode => {
      const nodes = ([] as YamlUnistNode[]).concat(selectNode(root));
      nodes.forEach(node => {
        test(getTestTitle(text), () => {
          expect(wrap(snapshotNode(text, node, options))).toMatchSnapshot();
        });
      });
    });
  });
}

function testCrLf(lfRoot: Root, lfText: string) {
  const crLfText = lfText.replace(/\n/g, "\r\n");

  test(getTestTitle(lfText), () => {
    const crLfRoot = parse(crLfText);
    testNode(lfRoot, crLfRoot);
  });

  function testNode(lfNode: Node, crLfNode: Node) {
    const [lfNodeProps, lfValueProps] = groupProps(lfNode);
    const [crLfNodeProps, crLfValueProps] = groupProps(crLfNode);

    expect(lfNodeProps).toEqual(crLfNodeProps);
    expect(lfValueProps).toEqual(crLfValueProps);

    for (const prop of lfValueProps as Array<keyof Node>) {
      const lfValue = lfNode[prop];
      const crLfValue = crLfNode[prop];

      if (typeof crLfValue === "string") {
        expect(crLfValue.replace(/\r\n/g, "\n")).toEqual(lfValue);
      } else {
        expect(crLfValue).toEqual(lfValue);
      }
    }

    for (const prop of lfNodeProps as Array<keyof Node>) {
      const lfChildNode = lfNode[prop];
      const crLfChildNode = crLfNode[prop];

      if (Array.isArray(lfChildNode) && Array.isArray(crLfChildNode)) {
        for (let i = 0; i < lfChildNode.length; i++) {
          testNode(lfChildNode[i], crLfChildNode[i]);
        }
      } else {
        expect(isNode(lfChildNode)).toBe(true);
        expect(isNode(crLfChildNode)).toBe(true);
        testNode(lfChildNode as Node, crLfChildNode as Node);
      }
    }

    expect([
      crLfNode.type,
      JSON.stringify(
        getOriginalText(crLfNode, crLfText).replace(/\r\n/g, "\n"),
      ),
    ]).toEqual([lfNode.type, JSON.stringify(getOriginalText(lfNode, lfText))]);
  }

  function getOriginalText(node: Node, text: string) {
    return text.slice(node.position.start.offset, node.position.end.offset);
  }

  function groupProps(node: Node) {
    const nodeProps: string[] = [];
    const valueProps: string[] = [];

    for (const key of Object.keys(node) as Array<keyof Node>) {
      if (key === "position") {
        continue;
      }

      const value = node[key];

      if (isNode(value) || (Array.isArray(value) && value.some(isNode))) {
        nodeProps.push(key);
      } else {
        valueProps.push(key);
      }
    }

    return [nodeProps, valueProps];
  }

  function isNode(value: any): value is Node {
    return value ? typeof value.type === "string" : false;
  }
}

function getTestTitle(text: string) {
  return JSON.stringify(text).slice(0, 60);
}

function getNodeDescription(node: Exclude<YamlUnistNode, null>) {
  return `${node.type} (${[
    `${node.position.start.line}:${node.position.start.column}`,
    `${node.position.end.line}:${node.position.end.column}`,
  ].join(" ~ ")})`;
}

function snapshotNode(
  text: string,
  node: YamlUnistNode,
  options: SnapshotNodeOptions = {},
) {
  if (node === null) {
    return "null";
  }

  const { selectNodeToStringify = (x: YamlUnistNode) => x } = options;
  const stringifiedNode = selectNodeToStringify(node)!;
  return (
    `${getNodeDescription(node)}\n` +
    codeFrameColumns(text, node.position, options.codeFrameMaxHeight) +
    "\n" +
    (node !== stringifiedNode
      ? `${getNodeDescription(stringifiedNode)}\n`
      : "") +
    stringifyNode(stringifiedNode, options)
  ).replace(/ +$/gm, "");
}

export interface StringifyNodeOptions {
  maxChildrenLevel?: number;
  maxCommentsLevel?: number;
}

function stringifyNode(
  node: YamlUnistNode | Node,
  options: StringifyNodeOptions = {},
): string {
  if (node === null) {
    return "<null />";
  }

  const attributes = Object.keys(node)
    .filter(key => {
      switch (key) {
        case "type":
        case "position":
        case "children":
        case "comments":
        case "leadingComments":
        case "middleComments":
        case "trailingComment":
        case "endComments":
        case "indicatorComment":
        case "anchor":
        case "tag":
          return false;
        default:
          return true;
      }
    })
    .map(key => `${key}=${JSON.stringify(node[key as keyof typeof node])}`)
    .sort()
    .map((attribute, index) => (index === 0 ? " " + attribute : attribute))
    .join(" ");
  const propNodes =
    "tag" in node
      ? (
          [node.tag, node.anchor].filter(
            propNode => propNode !== null,
          ) as Array<NonNullable<Tag | Anchor>>
        ).map(propNode => stringifyNode(propNode))
      : [];
  const comments =
    options.maxCommentsLevel === undefined || options.maxCommentsLevel > 0
      ? ([] as string[])
          .concat("leadingComments" in node ? "leadingComments" : [])
          .concat("middleComments" in node ? "middleComments" : [])
          .concat("indicatorComment" in node ? "indicatorComment" : [])
          .concat("trailingComment" in node ? "trailingComment" : [])
          .concat("endComments" in node ? "endComments" : [])
          .map(key =>
            // @ts-ignore
            ([].concat(node[key]).filter(Boolean) as Comment[]).map(
              comment =>
                `<${key.replace(/s$/, "")} value=${JSON.stringify(
                  comment.value,
                )}>`,
            ),
          )
          .reduce((a, b) => a.concat(b), [])
      : [];
  return "children" in node || comments.length !== 0 || propNodes.length !== 0
    ? `<${node.type}${attributes}>\n${indent(
        ([] as string[])
          .concat(
            propNodes,
            comments,
            (options.maxChildrenLevel === undefined ||
              options.maxChildrenLevel > 0) &&
              "children" in node
              ? (node.children as YamlUnistNode[]).map(childNode =>
                  stringifyNode(childNode, {
                    ...options,
                    ...(options.maxChildrenLevel !== undefined && {
                      maxChildrenLevel: options.maxChildrenLevel - 1,
                    }),
                    ...(options.maxCommentsLevel !== undefined && {
                      maxCommentsLevel: options.maxCommentsLevel - 1,
                    }),
                  }),
                )
              : [],
          )
          .join("\n"),
      )}\n</${node.type}>`
    : `<${node.type}${attributes} />`;
}

function indent(text: string) {
  return text
    .split("\n")
    .map(line => "  " + line)
    .join("\n");
}

function codeFrameColumns(
  text: string,
  position: Position,
  codeFrameMaxHeight = Infinity,
) {
  const lines = text.split("\n").map(line => `${line}¶`);
  const markerLines = lines.map((line, index) => {
    if (index < position.start.line - 1 || index > position.end.line - 1) {
      return "";
    }
    if (index === position.start.line - 1) {
      return (
        " ".repeat(position.start.column - 1) +
        (index === position.end.line - 1 &&
        position.start.column === position.end.column
          ? "~"
          : "^".repeat(
              (index === position.end.line - 1
                ? position.end.column - 1
                : line.length) -
                (position.start.column - 1),
            ))
      );
    } else if (index === position.end.line - 1) {
      return position.end.column === 1 && line === "¶"
        ? "^"
        : "^".repeat(position.end.column - 1);
    } else {
      return "^".repeat(line.length);
    }
  });

  const start = Math.max(0, position.start.line - 1 - codeFrameMaxHeight);
  const end = Math.min(lines.length, position.end.line + codeFrameMaxHeight);

  const gutterWidth = Math.floor(Math.log10(lines.length)) + 1;

  return lines
    .slice(start, end)
    .reduce((reduced, line, index) => {
      const gutter = leftpad(`${index + 1 + start}`, gutterWidth);
      return reduced.concat(
        `${gutter} | ${line.replace(/ /g, "·")}`,
        markerLines[index + start]
          ? `${" ".repeat(gutterWidth)} | ${markerLines[index + start]}`
          : [],
      );
    }, [] as string[])
    .join("\n");
}

function leftpad(text: string, width: number) {
  return " ".repeat(Math.max(0, width - text.length)) + text;
}

export function testSyntaxError(text: string, message?: string) {
  try {
    parse(text);
    throw new Error("SyntaxError not found");
  } catch (error: any) {
    if (!isYAMLError(error)) {
      throw error;
    }
    test(message || error.message, () => {
      expect(
        error.message + "\n" + codeFrameColumns(error.source, error.position),
      ).toMatchSnapshot();
    });
  }
}

function isYAMLError(
  e: any,
): e is YAML.YAMLSyntaxError | YAML.YAMLSemanticError {
  return (
    e instanceof Error &&
    (e.name === "YAMLSyntaxError" || e.name === "YAMLSemanticError")
  );
}
