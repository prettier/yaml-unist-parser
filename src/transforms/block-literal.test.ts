import { getFirstContent, testCases } from "../helpers.js";
import { BlockLiteral } from "../types.js";

testCases([
  ["|\n  123\n  456\n\n\n", getFirstContent()],
  ["|+\n  123\n  456\n\n\n", getFirstContent()],
  ["|-\n  123\n  456\n\n\n", getFirstContent()],
  ["|1\n  123\n  456\n\n\n", getFirstContent()],
  ["|1+\n  123\n  456\n\n\n", getFirstContent()],
  ["|1-\n  123\n  456\n\n\n", getFirstContent()],
  ["| # hello\n", root => root],
  ["| # hello \n  123\n  456\n\n\n", root => root],
  [
    "  !!str &anchor \n  # comment  \n|\n  123\n  456\n\n\n",
    [
      getFirstContent(),
      root => getFirstContent<BlockLiteral>(root).middleComments[0],
    ],
  ],
  ["!!str\n#middleComment\n| #indicatorComment\n value", getFirstContent()],
]);
