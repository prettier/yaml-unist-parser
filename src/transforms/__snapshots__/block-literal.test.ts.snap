// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`"  !!str &anchor \\n  # comment  \\n|\\n  123\\n  456\\n\\n\\n" 1`] = `
blockLiteral (3:1 ~ 6:1)
1 | ··!!str·&anchor·¶
2 | ··#·comment··¶
3 | |¶
  | ^^
4 | ··123¶
  | ^^^^^^
5 | ··456¶
  | ^^^^^^
6 | ¶
  | ^
7 | ¶
8 | ¶
<blockLiteral chomping="clip" indent=null value="123\\n456\\n">
  <tag value="tag:yaml.org,2002:str" />
  <anchor value="anchor" />
  <middleComment value=" comment  ">
</blockLiteral>
`;

exports[`"  !!str &anchor \\n  # comment  \\n|\\n  123\\n  456\\n\\n\\n" 2`] = `
comment (2:3 ~ 2:14)
1 | ··!!str·&anchor·¶
2 | ··#·comment··¶
  |   ^^^^^^^^^^^
3 | |¶
4 | ··123¶
5 | ··456¶
6 | ¶
7 | ¶
8 | ¶
<comment value=" comment  " />
`;

exports[`"!!str\\n#middleComment\\n| #indicatorComment\\n value" 1`] = `
blockLiteral (3:1 ~ 4:7)
1 | !!str¶
2 | #middleComment¶
3 | |·#indicatorComment¶
  | ^^^^^^^^^^^^^^^^^^^^
4 | ·value¶
  | ^^^^^^
<blockLiteral chomping="clip" indent=null value="value\\n">
  <tag value="tag:yaml.org,2002:str" />
  <middleComment value="middleComment">
  <indicatorComment value="indicatorComment">
</blockLiteral>
`;

exports[`"| # hello \\n  123\\n  456\\n\\n\\n" 1`] = `
root (1:1 ~ 6:1)
1 | |·#·hello·¶
  | ^^^^^^^^^^^
2 | ··123¶
  | ^^^^^^
3 | ··456¶
  | ^^^^^^
4 | ¶
  | ^
5 | ¶
  | ^
6 | ¶
  | ^
<root>
  <document>
    <documentHead>

    </documentHead>
    <documentBody>
      <blockLiteral chomping="clip" indent=null value="123\\n456\\n">
        <indicatorComment value=" hello ">
      </blockLiteral>
    </documentBody>
  </document>
</root>
`;

exports[`"| # hello\\n" 1`] = `
root (1:1 ~ 2:1)
1 | |·#·hello¶
  | ^^^^^^^^^^
2 | ¶
  | ^
<root>
  <document>
    <documentHead>

    </documentHead>
    <documentBody>
      <blockLiteral chomping="clip" indent=null value="">
        <indicatorComment value=" hello">
      </blockLiteral>
    </documentBody>
  </document>
</root>
`;

exports[`"|+\\n  123\\n  456\\n\\n\\n" 1`] = `
blockLiteral (1:1 ~ 6:1)
1 | |+¶
  | ^^^
2 | ··123¶
  | ^^^^^^
3 | ··456¶
  | ^^^^^^
4 | ¶
  | ^
5 | ¶
  | ^
6 | ¶
  | ^
<blockLiteral chomping="keep" indent=null value="123\\n456\\n\\n\\n" />
`;

exports[`"|\\n  123\\n  456\\n\\n\\n" 1`] = `
blockLiteral (1:1 ~ 4:1)
1 | |¶
  | ^^
2 | ··123¶
  | ^^^^^^
3 | ··456¶
  | ^^^^^^
4 | ¶
  | ^
5 | ¶
6 | ¶
<blockLiteral chomping="clip" indent=null value="123\\n456\\n" />
`;

exports[`"|-\\n  123\\n  456\\n\\n\\n" 1`] = `
blockLiteral (1:1 ~ 4:1)
1 | |-¶
  | ^^^
2 | ··123¶
  | ^^^^^^
3 | ··456¶
  | ^^^^^^
4 | ¶
  | ^
5 | ¶
6 | ¶
<blockLiteral chomping="strip" indent=null value="123\\n456" />
`;

exports[`"|1+\\n  123\\n  456\\n\\n\\n" 1`] = `
blockLiteral (1:1 ~ 6:1)
1 | |1+¶
  | ^^^^
2 | ··123¶
  | ^^^^^^
3 | ··456¶
  | ^^^^^^
4 | ¶
  | ^
5 | ¶
  | ^
6 | ¶
  | ^
<blockLiteral chomping="keep" indent=1 value="  123\\n  456\\n\\n\\n" />
`;

exports[`"|1\\n  123\\n  456\\n\\n\\n" 1`] = `
blockLiteral (1:1 ~ 4:1)
1 | |1¶
  | ^^^
2 | ··123¶
  | ^^^^^^
3 | ··456¶
  | ^^^^^^
4 | ¶
  | ^
5 | ¶
6 | ¶
<blockLiteral chomping="clip" indent=1 value="  123\\n  456\\n" />
`;

exports[`"|1-\\n  123\\n  456\\n\\n\\n" 1`] = `
blockLiteral (1:1 ~ 4:1)
1 | |1-¶
  | ^^^^
2 | ··123¶
  | ^^^^^^
3 | ··456¶
  | ^^^^^^
4 | ¶
  | ^
5 | ¶
6 | ¶
<blockLiteral chomping="strip" indent=1 value="  123\\n  456" />
`;
